phonenumbers.go

```diff
@@ -3412,20 +3410,41 @@ func GetTimezonesForNumber(number *PhoneNumber) ([]string, error) {
 	return GetTimezonesForPrefix(e164)
 }
 
-func getValueForNumber(onceMap map[string]*sync.Once, langMap map[string]*intStringMap, binMap map[string]string, language string, maxLength int, number *PhoneNumber) (string, int, error) {
-	// do we have data for this language
-	_, existing := binMap[language]
-	if !existing {
-		return "", 0, nil
-	}
+func fillLangMap(langMap map[string]*intStringMap, file fs.File, language string) error {
+	m.Lock()
+	defer m.Unlock()
 
 	// load it into our map
-	onceMap[language].Do(func() {
-		prefixMap, err := loadPrefixMap(binMap[language])
-		if err == nil {
-			langMap[language] = prefixMap
+	data, err := io.ReadAll(file)
+	if err != nil {
+		return err
+	}
+
+	prefixMap, err := loadPrefixMap(data)
+	if err == nil {
+		langMap[language] = prefixMap
+	}
+
+	return nil
+}
+
+func getValueForNumber(langMap map[string]*intStringMap, dir string, binMap embed.FS, language string, maxLength int, number *PhoneNumber) (string, int, error) {
+	// do we have data for this language
+	file, err := binMap.Open(dir + "/" + language + ".txt.gz")
+	if err != nil {
+		if errors.Is(err, fs.ErrNotExist) {
+			return "", 0, nil
 		}
-	})
+
+		return "", 0, err
+	}
+
+	defer file.Close()
```

Comment:
Why is the file existence check always repeated? Remember package initialization isn't the only concern here - lookup times are more important.

I think `getValueForNumber` should work like...

1. inside mutex...
2. check if `langMap` has value for `lang`
3. try to load from file if not, storing result in `langMap`
4. use value from `langMap`

Reply:
yes, i was wondering the same the old code checked whether the language key exists in the metadata/binMap first and then it parses the file...

Reply:
@rowanseymour sorry i am making the changes but, do you know why we do not return an error when the field do not exist?

Reply:
ok, I applied the changes I am not sure it is according to your idea, let me know if you see how it can be improved

I added a benchmark these are the results:

prev change:
BenchmarkGetCarrierWithPrefixForNumber 	      30	  34821182 ns/op	 2948078 B/op	   28948 allocs/op

main
BenchmarkGetCarrierWithPrefixForNumber 	  109204	     11091 ns/op	     401 B/op	       7 allocs/op

new
BenchmarkGetCarrierWithPrefixForNumber 	  107184	     11363 ns/op	     448 B/op	       8 allocs/op

Reply:
That benchmark tho is just checking the same language over and over again which means only one data load.. but I think it's ok